---
title: "Analysis"
output: html_notebook
---


```{r, Load}
library(survival)
library(survminer)
library(dplyr)
source("run_analysis_pipeline.R")
source("PanClockHelperFunctions.R")
```

# Runs CoxPH given experiment and predictions and output object containing each CoxPH run
```{r, GetSurvivalStats}
getSurvivalStats <- function(fullExperiment, predictionObject, applyBiasCorrection=FALSE, useGender=TRUE, useCovariates=TRUE, degreesFreedomNonLinear=8, useComplicatedNonlinear=TRUE){
  
  if (applyBiasCorrection) { predictedAges <- predictionObject$predicted_corrected_age }
  else { predictedAges <- predictionObject$predicted_age }
  
  prediction_df <- data.frame(predicted_age=predictedAges, submitter_id=predictionObject$submitter_id)
  survivalOutput <- run_analysis_pipeline(fullExperiment, prediction_df, useCovariates=useCovariates, degreesFreedomNonLinear=degreesFreedomNonLinear, useComplicatedNonlinear=useComplicatedNonlinear)
  
  survivalOutput$non_interaction_linear_model$compareTo <- "baseline_model"
  survivalOutput$non_interaction_nonlinear_model$compareTo <- "baseline_nonlinear_model"
  survivalOutput$interaction_linear_model$compareTo <- "baseline_interaction_model"
  survivalOutput$interaction_linear_duplicate_model$compareTo <- "baseline_interaction2_model"
  survivalOutput$interaction_additive_nonlinear_model$compareTo <- "baseline_interaction_nonlinear_additive_model"
  return(survivalOutput)
}
```

# Gets likelihood ratio between two models
```{r, GetTrueLikelihoodRatio}
getTrueLikelihoodRatio <- function(baselineModel, modifiedModel){
  
  baselineLikelihood <- logLik(baselineModel)
  modifiedLikelihood <- logLik(modifiedModel)
  newLogRatio <- -2 * (baselineLikelihood - modifiedLikelihood)
  degreesOfFreedom <- length(coef(modifiedModel)) - length(coef(baselineModel))
  pValue <- 1 - pchisq(newLogRatio, degreesOfFreedom)
  return(list(logratio=newLogRatio, pval=pValue))
}
```

### Save age model features and summary stats as well as survival analysis. Important to set additionalDescriptor to unique name if you don't want to overwrite previous outputs
```{r, SaveClockOutput}
saveClockOutput <- function(cancerType, layers, modelOutput, dir, additionalDescriptor, crossValidating=FALSE, applyBiasCorrection=FALSE){
  
  # Modify output path
  cancerLayersPrefix <- getCancerLayersPrefix(cancerType, layers)
  primaryDir <- paste0(dir, cancerType)
  dir.create(primaryDir, recursive = TRUE, showWarnings = FALSE)
  filePath <- paste0(primaryDir, "/", cancerLayersPrefix)

  # Get coefficients
  coefficientFrame <- list()
  if (!crossValidating){
    coefficientFrame[[cancerLayersPrefix]] <- getCoefficientDF(modelOutput$ModelBuilding$model$finalModel)
  } else{
     coefficientFrame[[cancerLayersPrefix]] <- modelOutput$ModelBuilding$weights
  }
  
  # # Get stats
  # stats <- getSurvivalStats(modelOutput$experimentList[[1]], modelOutput$ModelBuilding$predicted, applyBiasCorrection=applyBiasCorrection)
  # statsTable <- getFormattedStatsTable(stats)
  # summaryTable <- getSummaryTable(cancerType, layers, modelOutput, stats)
  # print(statsTable)
  
  # Save outputs
  coefficientPath <- paste0(filePath, "_coef", additionalDescriptor, ".rds")
  saveRDS(coefficientFrame, file = coefficientPath)
  # write.csv(summaryTable, paste0(filePath, "_summary", additionalDescriptor, ".csv"), row.names = FALSE)
  # write.csv(statsTable, paste0(filePath, "_stats", additionalDescriptor, ".csv"), row.names = TRUE)
  return(list(primaryDir=primaryDir, coefficientPath=coefficientPath))
}
```

### Perform pathway analysis
```{r, PerformEnrichment}
performEnrichment <- function(cancerType, layers, additionalDescriptor, clockOutput){
  
  # Initialize files and folders
  cancerLayersPrefix <- getCancerLayersPrefix(cancerType, layers)
  outputDir=paste0(clockOutput$primaryDir,"/Enrichment/", additionalDescriptor)
  dir.create(outputDir, recursive = TRUE, showWarnings = FALSE)
  formattedWeightFileName <- paste0(outputDir, "/", cancerLayersPrefix, "_weights_named.csv") 

  # Run pathway analysis
  process_gene_weights(cancerType, weights_file=clockOutput$coefficientPath, output_dir=outputDir)
  run_pathway_analysis(cancerType, input_file=formattedWeightFileName, output_dir=outputDir)
}
```

### Get coefficient df
```{r, GetCoefficientDF}
getCoefficientDF <- function(finalModel){

  # Extract lambda values and coefficients
  lambda_values <- finalModel$lambda
  closest_lambda_index <- which.min(abs(lambda_values - finalModel$lambdaOpt))
  beta_matrix <- finalModel$beta
  coef_names <- rownames(beta_matrix)
  beta_column <- beta_matrix[, closest_lambda_index]

  coef_df <- data.frame(
    Feature = coef_names,
    Weight = as.numeric(beta_column)
  )
  rownames(coef_df) <- coef_df$Feature
  return(coef_df)
}
```

### Map RNAseq gene IDs to names
```{r, GetGeneMap}
getGeneMap <- function(modelOutput){
  features <- rownames(assay(modelOutput$experimentList[[1]], "unstranded"))
  geneIDs <- c()
  geneNames <- c()
  RNAseqExperiment <- modelOutput$experimentList$RNAseq
  for (name in features){
    if (substr(name, 1, 4) == "ENSG"){
      geneIDs <- c(geneIDs, name)
      geneIndex <- which(rowRanges(RNAseqExperiment)$gene_id %in% name)
      geneName <- rowRanges(RNAseqExperiment)$gene_name[geneIndex]
      geneNames <- c(geneNames, geneName)
    }
  }
  
  geneMap <- data.frame(Ensembl_ID=geneIDs, Gene_Name=geneNames)
  return(geneMap)
}

```

### Find coefficients of individual features used in final model
```{r, OrganizeOutputFeatures}
## For RNAseq data
# To find coefficients: coefficients <- coef(modelPredictionsPair$model$finalModel, modelPredictionsPair$model$bestTune$lambda)
# To find the index of a coefficient: index <- which(coefficients@x %in% COEFFICIENT_OF_CHOICE)
# To find the gene ID of an index: geneID <- coefficients@Dimnames[[1]][index]
# To find the index of the gene in the full experiment: fullIndex <- which(rowRanges(experimentsList[[1]])$gene_id %in% geneID) 
# To find the gene name: featureName <- rowRanges(experimentsList[[1]])$gene_name[fullIndex]
# RNASeqElNet <- read.csv("./ModelBuildingOutput/RNAseqElNet.csv")

organizeOutputFeatures <- function(modelOutput, RNAExperiment=NULL){

  coefficients <- coef(modelOutput$model$finalModel, modelOutput$model$bestTune$lambda)
  # coefficients2 <- coef(modelOutput$model$finalModel, unlist(modelOutput$model$bestTune))
  featureFrame <- data.frame(features=coefficients@Dimnames[[1]][coefficients@i + 1], coefficients=coefficients@x)
  featureNames <- c()
  origin <- c()
  for (name in featureFrame$features){
    featureIdx <- which(colnames(modelOutput$predicted) %in% name)
    if (!is.null(RNAExperiment) && substr(name, 1, 4) == "ENSG"){
      origin <- c(origin, "RNAseq")
      geneIndex <- which(rowRanges(experimentsList[[1]])$gene_id %in% name)
      geneName <- rowRanges(RNAExperiment)$gene_name[geneIndex]
      featureNames <- append(featureNames, rowRanges(modelOutput$experimentsList$RNAseq)$gene_name[geneIndex])
    }
    else{
      featureNames <- append(featureNames, name)
      if(substr(name, 2, 4) == "hsa"){origin <- c(origin, "miRNA")}
      else if(length(featureIdx) > 0 && featureIdx < 9){origin <- c(origin, "covariate")}
      else if(name == "(Intercept)"){origin <- c(origin, "Intercept")}
      else{origin <- c(origin, "RPPA")}
    }
  }
  
  featureFrame$names <- featureNames
  featureFrame$origin <- origin
  featureFrame$absCoefficients <- abs(featureFrame$coefficients)
  featureFrame <- featureFrame[order(featureFrame$absCoefficients, decreasing=TRUE),]
  return(featureFrame)
}

# featureFrame <- organizeOutputFeatures(modelPredictionsPair, RNAExperiment=experimentsList[["RNAseq"]])
# 
# outFile <- "./ModelBuildingOutput/ElasticNetRNAseqHoldoutsFeatures.csv"
# write.csv(featureFrame, outFile, row.names = FALSE)
```

```{r, IntializeSummaryTable}
initializeSummaryTable <- function(){
  cancer_summary_table <- data.frame(
    test_name = character(),
    cancer_type = character(),
    model = character(),
    age_association_adjusted_p_cutoff = numeric(),
    layer_combination = character(),
    
    Rsquared = numeric(),
    RMSE = numeric(),
    alpha = numeric(),
    lambda = numeric(),
    non_zero_weight_feature_count = numeric(),
    zero_weight_feature_count = numeric(),
    
    interaction_against_baseline_likelihood_ratio = numeric(),
    interaction_against_baseline_pval = numeric(),
    interaction_against_baseline_adjusted_pval = numeric(),
    interaction_against_non_interaction_likelihood_ratio = numeric(),
    interaction_against_non_interaction_pval = numeric(),
    interaction_against_non_interaction_adjusted_pval = numeric(),
    interaction_term_coeff_interaction = numeric(),
    interaction_term_exp_coeff_interaction = numeric(),
    interaction_term_pval_interaction = numeric(),
    interaction_term_adjusted_pval_interaction = numeric(),
    delta_age_coeff_interaction	= numeric(),
    delta_age_exp_coeff_interaction = numeric(),
    delta_age_pval_interaction = numeric(),
    delta_age_adjusted_pval_interaction = numeric(),
    chronological_coef_interaction = numeric(),
    chronological_exp_coef_interaction = numeric(),
    chronological_pval_interaction = numeric(),
    Concordance_interaction = numeric(),
    se_concordance_interaction = numeric(),
    lik_ratio_test_interaction = numeric(),
    lik_ratio_test_df_interaction = numeric(),
    lik_ratio_test_pval_interaction = numeric(),
    Wald_test_interaction = numeric(),
    Wald_test_df_interaction = numeric(),
    Wald_test_pval_interaction = numeric(),
    logrank_score_test_interaction = numeric(),
    logrank_score_test_df_interaction = numeric(),
    logrank_score_test_pval_interaction = numeric(),
    
    non_interaction_against_baseline_likelihood_ratio = numeric(),
    non_interaction_against_baseline_adjusted_pval = numeric(),
    delta_age_coeff_non_interaction	= numeric(),
    delta_age_exp_coeff_non_interaction = numeric(),
    delta_age_pval_non_interaction = numeric(),
    delta_age_adjusted_pval_non_interaction = numeric(),
    chronological_coef_non_interaction = numeric(),
    chronological_exp_coef_non_interaction = numeric(),
    chronological_pval_non_interaction = numeric(),
    Concordance_non_interaction = numeric(),
    se_concordance_non_interaction = numeric(),
    lik_ratio_test_non_interaction = numeric(),
    lik_ratio_test_df_non_interaction = numeric(),
    lik_ratio_test_pval_non_interaction = numeric(),
    Wald_test_non_interaction = numeric(),
    Wald_test_df_non_interaction = numeric(),
    Wald_test_pval_non_interaction = numeric(),
    logrank_score_test_non_interaction = numeric(),
    logrank_score_test_df_non_interaction = numeric(),
    logrank_score_test_pval_non_interaction = numeric(),
    
    chronological_coef_baseline = numeric(),
    chronological_exp_coef_baseline = numeric(),
    chronological_pval_baseline = numeric(),
    chronological_adjusted_pval_baseline = numeric(),
    Concordance_baseline = numeric(),
    se_concordance_baseline = numeric(),
    lik_ratio_test_baseline = numeric(),
    lik_ratio_test_df_baseline = numeric(),
    lik_ratio_test_pval_baseline = numeric(),
    Wald_test_baseline = numeric(),
    Wald_test_df_baseline = numeric(),
    Wald_test_pval_baseline = numeric(),
    logrank_score_test_baseline = numeric(),
    logrank_score_test_df_baseline = numeric(),
    logrank_score_test_pval_baseline = numeric(),
    
    predicted_age_coef_biological = numeric(),
    predicted_age_exp_coef_biological = numeric(),
    predicted_age_pval_biological = numeric(),
    predicted_age_adjusted_pval_biological = numeric(),
    Concordance_biological = numeric(),
    se_concordance_biological = numeric(),
    lik_ratio_test_biological = numeric(),
    lik_ratio_test_df_biological = numeric(),
    lik_ratio_test_pval_biological = numeric(),
    Wald_test_biological = numeric(),
    Wald_test_df_biological = numeric(),
    Wald_test_pval_biological = numeric(),
    logrank_score_test_biological = numeric(),
    logrank_score_test_df_biological = numeric(),
    logrank_score_test_pval_biological = numeric()
  )
  return(cancer_summary_table)
}
```

```{r, GetSummaryTable}
getSummaryTable <- function(comb, modelOutput, stats, params=NULL){
  interactionLikelihood <- getTrueLikelihoodRatio(stats$baseline_model, stats$interaction_model)
  nonInteractionLikelihood <- getTrueLikelihoodRatio(stats$baseline_model, stats$non_interaction_model)
  interactionNonInteractionLikelihood <- getTrueLikelihoodRatio(stats$non_interaction_model, stats$interaction_model)
  if ("Fold1" %in% names(modelOutput$ModelBuilding$model) || "Fold01" %in% names(modelOutput$ModelBuilding$model)){
    alpha <- modelOutput$ModelBuilding$model[[1]]$model$bestTune$alpha
    lambda <- modelOutput$ModelBuilding$model[[1]]$model$bestTune$lambda
  }
  else{
    alpha = modelOutput$ModelBuilding$model$bestTune$alpha
    lambda = modelOutput$ModelBuilding$model$bestTune$lambda
  }
  non_zero_weights <- sum(modelOutput$ModelBuilding$weights$Weight != 0, na.rm = TRUE)
  zero_weights <- sum(modelOutput$ModelBuilding$weights$Weight == 0, na.rm = TRUE)
  if (!is.null(params)){
    test_name <- params$test_name
    model <- params$model_type
    age_association_adjusted_p_cutoff <- params$significance_cutoff
    cancer_type <- params$cancer_type
  }
  else{
    test_name <- "None"
    model <- "None"
    age_association_adjusted_p_cutoff <- "None"
    cancer_type <- "None"
  }

  summaryTable <- data.frame(
                    test_name = test_name,
                    cancer_type = cancer_type,
                    model = model,
                    age_association_adjusted_p_cutoff = age_association_adjusted_p_cutoff,
                    layer_combination = paste(comb, collapse = "_"),
                    
                    Rsquared = modelOutput$ModelBuilding$model$R2,
                    RMSE = modelOutput$ModelBuilding$model$RMSE,
                    alpha = alpha,
                    lambda = lambda,
                    non_zero_weight_feature_count = non_zero_weights,
                    zero_weight_feature_count = zero_weights,
                    
                    interaction_against_baseline_likelihood_ratio = interactionLikelihood$logratio,
                    interaction_against_baseline_pval = interactionLikelihood$pval,
                    interaction_against_baseline_adjusted_pval = interactionLikelihood$pval,
                    interaction_against_non_interaction_likelihood_ratio =interactionNonInteractionLikelihood$logratio,
                    interaction_against_non_interaction_pval = interactionNonInteractionLikelihood$pval,
                    interaction_against_non_interaction_adjusted_pval = interactionLikelihood$pval,
                    interaction_term_coeff_interaction = stats$interaction_summary$coefficients["chronological:delta_age", "coef"],
                    interaction_term_exp_coeff_interaction = stats$interaction_summary$coefficients["chronological:delta_age", "exp(coef)"],
                    interaction_term_pval_interaction = stats$interaction_summary$coefficients["chronological:delta_age", "Pr(>|z|)"],
                    interaction_term_adjusted_pval_interaction = stats$interaction_summary$coefficients["chronological:delta_age", "Pr(>|z|)"],
                    delta_age_coeff_interaction = stats$interaction_summary$coefficients["delta_age", "coef"],
                    delta_age_exp_coeff_interaction = stats$interaction_summary$coefficients["delta_age", "exp(coef)"],
                    delta_age_pval_interaction = stats$interaction_summary$coefficients["delta_age", "Pr(>|z|)"],
                    delta_age_adjusted_pval_interaction = stats$interaction_summary$coefficients["delta_age", "Pr(>|z|)"],
                    chronological_coef_interaction = stats$interaction_summary$coefficients["chronological", "coef"],
                    chronological_exp_coef_interaction = stats$interaction_summary$coefficients["chronological", "exp(coef)"],
                    chronological_pval_interaction = stats$interaction_summary$coefficients["chronological", "Pr(>|z|)"],
                    Concordance_interaction = stats$interaction_summary$concordance[1],
                    se_concordance_interaction = stats$interaction_summary$concordance[2],
                    lik_ratio_test_interaction = stats$interaction_summary$logtest["test"],
                    lik_ratio_test_df_interaction = stats$interaction_summary$logtest["df"],
                    lik_ratio_test_pval_interaction = stats$interaction_summary$logtest["pvalue"],
                    Wald_test_interaction = stats$interaction_summary$waldtest["test"],
                    Wald_test_df_interaction = stats$interaction_summary$waldtest["df"],
                    Wald_test_pval_interaction = stats$interaction_summary$waldtest["pvalue"],
                    logrank_score_test_interaction = stats$interaction_summary$sctest["test"],
                    logrank_score_test_df_interaction = stats$interaction_summary$sctest["df"],
                    logrank_score_test_pval_interaction = stats$interaction_summary$sctest["pvalue"],
                    
                    non_interaction_against_baseline_likelihood_ratio = nonInteractionLikelihood$logratio,
                    non_interaction_against_baseline_pval = nonInteractionLikelihood$pval,
                    non_interaction_against_baseline_adjusted_pval = nonInteractionLikelihood$pval,
                    delta_age_coeff_non_interaction	= stats$non_interaction_summary$coefficients["delta_age", "coef"],
                    delta_age_exp_coeff_non_interaction = stats$non_interaction_summary$coefficients["delta_age", "exp(coef)"],
                    delta_age_pval_non_interaction = stats$non_interaction_summary$coefficients["delta_age", "Pr(>|z|)"],
                    delta_age_adjusted_pval_non_interaction = stats$non_interaction_summary$coefficients["delta_age", "Pr(>|z|)"],
                    chronological_coef_non_interaction = stats$non_interaction_summary$coefficients["chronological", "coef"],
                    chronological_exp_coef_non_interaction = stats$non_interaction_summary$coefficients["chronological", "exp(coef)"],
                    chronological_pval_non_interaction =  stats$non_interaction_summary$coefficients["chronological", "Pr(>|z|)"],
                    Concordance_non_interaction = stats$non_interaction_summary$concordance[1],
                    se_concordance_non_interaction = stats$non_interaction_summary$concordance[2],
                    lik_ratio_test_non_interaction = stats$non_interaction_summary$logtest["test"],
                    lik_ratio_test_df_non_interaction = stats$non_interaction_summary$logtest["df"],
                    lik_ratio_test_pval_non_interaction = stats$non_interaction_summary$logtest["pvalue"],
                    Wald_test_non_interaction = stats$non_interaction_summary$waldtest["test"],
                    Wald_test_df_non_interaction = stats$non_interaction_summary$waldtest["df"],
                    Wald_test_pval_non_interaction = stats$non_interaction_summary$waldtest["pvalue"],
                    logrank_score_test_non_interaction = stats$non_interaction_summary$sctest["test"],
                    logrank_score_test_df_non_interaction = stats$non_interaction_summary$sctest["df"],
                    logrank_score_test_pval_non_interaction = stats$non_interaction_summary$sctest["pvalue"],
                    
                    chronological_coef_baseline = stats$baseline_summary$coefficients["chronological", "coef"],
                    chronological_exp_coef_baseline = stats$baseline_summary$coefficients["chronological", "exp(coef)"],
                    chronological_pval_baseline = stats$baseline_summary$coefficients["chronological", "Pr(>|z|)"],
                    chronological_adjusted_pval_baseline = stats$baseline_summary$coefficients["chronological", "Pr(>|z|)"],
                    Concordance_baseline = stats$baseline_summary$concordance[1],
                    se_concordance_baseline = stats$baseline_summary$concordance[2],
                    lik_ratio_test_baseline = stats$baseline_summary$logtest["test"],
                    lik_ratio_test_df_baseline = stats$baseline_summary$logtest["df"],
                    lik_ratio_test_pval_baseline = stats$baseline_summary$logtest["pvalue"],
                    Wald_test_baseline = stats$baseline_summary$waldtest["test"],
                    Wald_test_df_baseline = stats$baseline_summary$waldtest["df"],
                    Wald_test_pval_baseline = stats$baseline_summary$waldtest["pvalue"],
                    logrank_score_test_baseline = stats$baseline_summary$sctest["test"],
                    logrank_score_test_df_baseline = stats$baseline_summary$sctest["df"],
                    logrank_score_test_pval_baseline = stats$baseline_summary$sctest["pvalue"],
                    
                    predicted_age_coef_biological = stats$biological_summary$coefficients["predicted", "coef"],
                    predicted_age_exp_coef_biological = stats$biological_summary$coefficients["predicted", "exp(coef)"],
                    predicted_age_pval_biological = stats$biological_summary$coefficients["predicted", "Pr(>|z|)"],
                    predicted_age_adjusted_pval_biological = stats$biological_summary$coefficients["predicted", "Pr(>|z|)"],
                    Concordance_biological = stats$biological_summary$concordance[1],
                    se_concordance_biological = stats$biological_summary$concordance[2],
                    lik_ratio_test_biological = stats$biological_summary$logtest["test"],
                    lik_ratio_test_df_biological = stats$biological_summary$logtest["df"],
                    lik_ratio_test_pval_biological = stats$biological_summary$logtest["pvalue"],
                    Wald_test_biological = stats$biological_summary$waldtest["test"],
                    Wald_test_df_biological = stats$biological_summary$waldtest["df"],
                    Wald_test_pval_biological = stats$biological_summary$waldtest["pvalue"],
                    logrank_score_test_biological = stats$biological_summary$sctest["test"],
                    logrank_score_test_df_biological = stats$biological_summary$sctest["df"],
                    logrank_score_test_pval_biological = stats$biological_summary$sctest["pvalue"]
  )

  proportionalityAssumptionTestInteraction <- cox.zph(stats$interaction_model)
  interactionRownames <- rownames(proportionalityAssumptionTestInteraction$table)
  for (i in 1:length(interactionRownames)){
    summaryTable[[paste0("prop_assump_interaction_", interactionRownames[i])]] <- proportionalityAssumptionTestInteraction$table[i, 3]
  }
  
  proportionalityAssumptionTestNonInteraction <- cox.zph(stats$non_interaction_model)
  nonInteractionRownames <- rownames(proportionalityAssumptionTestNonInteraction$table)
  for (i in 1:length(nonInteractionRownames)){
    summaryTable[[paste0("prop_assump_non_interaction_", nonInteractionRownames[i])]] <- proportionalityAssumptionTestNonInteraction$table[i, 3]
  }
  
  proportionalityAssumptionTestBaseline <- cox.zph(stats$baseline_model)
  baselineRownames <- rownames(proportionalityAssumptionTestBaseline$table)
  for (i in 1:length(baselineRownames)){
    summaryTable[[paste0("prop_assump_baseline_", baselineRownames[i])]] <- proportionalityAssumptionTestBaseline$table[i, 3]
  }
  
  return(summaryTable)
}
  
```

# Get output table of important statistical findings
```{r, GetFormattedStatsTable}
getFormattedStatsTable <- function(modelOutput, comb, params=NULL, applyBiasCorrection=FALSE, useCovariates=FALSE, degreesFreedomNonLinear=4, useComplicatedNonlinear=FALSE){
  
  stats <- getSurvivalStats(modelOutput$experimentList[[1]], modelOutput$ModelBuilding$predicted, applyBiasCorrection=applyBiasCorrection, useCovariates=useCovariates, degreesFreedomNonLinear=degreesFreedomNonLinear, useComplicatedNonlinear=useComplicatedNonlinear)

  if ("Fold1" %in% names(modelOutput$ModelBuilding$model) || "Fold01" %in% names(modelOutput$ModelBuilding$model)){
    alpha <- modelOutput$ModelBuilding$model[[1]]$model$bestTune$alpha
    lambda <- modelOutput$ModelBuilding$model[[1]]$model$bestTune$lambda
  }
  else{
    alpha = modelOutput$ModelBuilding$model$bestTune$alpha
    lambda = modelOutput$ModelBuilding$model$bestTune$lambda
  }
  non_zero_weights <- sum(modelOutput$ModelBuilding$weights$Weight != 0, na.rm = TRUE)
  zero_weights <- sum(modelOutput$ModelBuilding$weights$Weight == 0, na.rm = TRUE)
  if (!is.null(params)){
    test_name <- params$test_name
    model <- params$model_type
    age_association_adjusted_p_cutoff <- params$significance_cutoff
    cancer_type <- params$cancer_type
  }
  else{
    test_name <- "None"
    model <- "None"
    age_association_adjusted_p_cutoff <- "None"
    cancer_type <- "None"
  }

  cancerLayerFrame <- data.frame(
                    test_name = test_name,
                    cancer_type = cancer_type,
                    layer_combination = paste(comb, collapse = "_"),
                    model = model,
                    age_association_adjusted_p_cutoff = age_association_adjusted_p_cutoff,
                    degreesFreedomNonLinear = degreesFreedomNonLinear,
                    
                    Rsquared = modelOutput$ModelBuilding$model$R2,
                    RMSE = modelOutput$ModelBuilding$model$RMSE,
                    alpha = alpha,
                    lambda = lambda,
                    non_zero_weight_feature_count = non_zero_weights,
                    zero_weight_feature_count = zero_weights
  )
                    
  for (name in names(stats)){
    if (grepl("model", name)){
      model <- stats[[name]]
      comparisonModel <- model$compareTo
      if (!is.null(comparisonModel)){
        trueLikelihoodRatio <- getTrueLikelihoodRatio(stats[[comparisonModel]], model)
        cancerLayerFrame[[paste0(name, "_likelihood_against_", comparisonModel)]] <- trueLikelihoodRatio$logratio
        cancerLayerFrame[[paste0(name, "_likelihood_against_", comparisonModel, "_pval")]] <- trueLikelihoodRatio$pval
      }

      summary <- summary(model)
      for (i in 1:length(summary$coefficients[, "coef"])){# rownames(as.data.frame(summary$coefficients))){
        coefficients <- summary$coefficients
        coefficient <- rownames(coefficients)[i]
        cancerLayerFrame[[paste0(name, "_", coefficient, "_weight")]] <- coefficients[i, "coef"]
        if ("p" %in% colnames(coefficients)){ cancerLayerFrame[[paste0(name, "_", coefficient, "_pval")]] <- coefficients[i, "p"] }
        else { cancerLayerFrame[[paste0(name, "_", coefficient, "_pval")]] <- coefficients[i, "Pr(>|z|)"] }
      }
      cancerLayerFrame[[paste0(name, "_likelikood_ratio")]] <- summary$logtest["test"]
      cancerLayerFrame[[paste0(name, "_likelikood_ratio_pval")]] <- summary$logtest["pvalue"]
      cancerLayerFrame[[paste0(name, "_concordance")]] <- summary$concordance[1]
    }
  }
  
  return(cancerLayerFrame)
}
```

# Function to produce a summary table of layers and cancers for their ML model info and survival stats. Excludes cancers with insufficient sample sizes.
```{r}
writeUltimateSummaryTable <- function(outputDir="/restricted/projectnb/agedisease/projects/pancancer_aging_clocks/results/Eitan/ComboRidge005/", suffix="_ridge_omics_summary.csv"){
  
  # Only make summary table for cancers of sufficient sample size, set cutoffs to 0 if you want to use all
  cancer_types <- getEligibleCancers("RNAseq", cutoff=250)
  cancer_types_methylation <- getEligibleCancers("methylation", cutoff=250)
  cancer_types <- cancer_types[cancer_types %in% cancer_types_methylation] 
  summary_file_paths <- paste0(outputDir, 
                        cancer_types, 
                        suffix)
  

  # Read and combine all CSVs
  summaries_list <- lapply(summary_file_paths, function(file) {
    if (file.exists(file)) {
      read.csv(file, row.names = NULL)  # Adjust row.names if needed
    } else {
      # warning(paste("File not found:", file))
      return(NULL)
    }
  })
  
  combined_tables <- bind_rows(summaries_list)

  if (!("cancer_type" %in% colnames(combined_tables))){
    rep_cancers <- rep(cancer_types, each=4)
    combined_tables <- cbind(cancer_type=rep_cancers, combined_tables)
  }
  
  # Multiple test correction for likelihood ratios
  combined_tables$interaction_against_baseline_adjusted_pval <- p.adjust(combined_tables$interaction_against_baseline_pval, method="fdr")
  combined_tables$non_interaction_against_baseline_adjusted_pval <- p.adjust(combined_tables$non_interaction_against_baseline_pval, method="fdr")
  combined_tables$interaction_against_non_interaction_adjusted_pval <- p.adjust(combined_tables$interaction_against_non_interaction_pval, method="fdr")
  
  # Multiple test correction for covariate significances
  combined_tables$interaction_term_adjusted_pval_interaction <- p.adjust(combined_tables$interaction_term_pval_interaction, method="fdr")
  combined_tables$delta_age_adjusted_pval_interaction <- p.adjust(combined_tables$delta_age_pval_interaction, method="fdr")
  combined_tables$delta_age_adjusted_pval_non_interaction <- p.adjust(combined_tables$delta_age_pval_non_interaction, method="fdr")
  combined_tables$chronological_adjusted_pval_baseline <- p.adjust(combined_tables$chronological_pval_baseline, method="fdr")
  combined_tables$predicted_age_adjusted_pval_biological <- p.adjust(combined_tables$predicted_age_pval_biological, method="fdr")
  
  # Export results
  write.csv(combined_tables, paste0(outputDir, "UltimateSummaryTable", suffix))
}

```

# Function to plot how the hazard ratio (representative of survival, potentially modulated by an interaction term) varies with age
```{r, HazardRatioFigure}
plotHazardRatio <- function(coxOutput, chronologicalAges, title, x="chronological", y="delta_age", usingInteractionTerm=TRUE, scale=1){
  # Extract coefficients and covariance matrix
  coef_vec <- coxOutput$coef
  cov_mat  <- coxOutput$var
  
  beta1_name <- y
  beta1 <- coef_vec[beta1_name]

  if (usingInteractionTerm){
    beta3_name <- paste0(x, ":", y)
    beta3 <- coef_vec[beta3_name]
  }
  else{
    beta3_name <- NULL
    beta3 <- 0
  }
  
  # --------------------------------------------
  # 2. Define the range of ages for the X-axis
  #    We'll take the min to max from dfSurv's data
  # --------------------------------------------
  min_age <- floor(min(chronologicalAges, na.rm=TRUE))
  max_age <- ceiling(max(chronologicalAges, na.rm=TRUE))
  ages_seq <- seq(min_age, max_age, by=1)
  
  # --------------------------------------------
  # 3. Compute log(HR) for a 1-unit increase
  #    = beta1 + beta3 * Age
  #    Then exponentiate to get HR
  # --------------------------------------------
  logHR <- scale * (beta1 + beta3 * ages_seq)
  HR    <- exp(logHR)
  
  # --------------------------------------------
  # 4. Compute a pointwise 95% CI for HR(Age)
  # --------------------------------------------
  # 1) Extract coefficient names 
  beta_names <- names(coxOutput$coef)
  
  # 2) Assign them as dimnames on the covariance matrix
  dimnames(cov_mat) <- list(beta_names, beta_names)
  
  var_b1 <- cov_mat[beta1_name, beta1_name]
  if (usingInteractionTerm){
    var_b3 <- cov_mat[beta3_name, beta3_name]
    cov_b1b3 <- cov_mat[beta1_name, beta3_name]
  }
  else{
    var_b3 <- 0
    cov_b1b3 <- 0
  }
  
  # Calculate variance of beta and (standard error)
  # For a linear function of two random variables (in this case, β₁ and β₃), the variance is calculated as:
  varBeta <- scale^2 * (var_b1 + ages_seq^2 * var_b3 + 2 * ages_seq * cov_b1b3)
  seBeta  <- sqrt(varBeta)
  
  logHR_low  <- logHR - 1.96 * seBeta
  logHR_high <- logHR + 1.96 * seBeta
  
  HR_low  <- exp(logHR_low)
  HR_high <- exp(logHR_high)
  
  # P data frame for plotting
  df_plot <- data.frame(
    age    = ages_seq,
    hr     = HR,
    hr_low = HR_low,
    hr_high= HR_high
  )
  
  # --------------------------------------------
  # 5. Create a data frame of per-sample points
  #    So each sample is plotted on the same line
  # --------------------------------------------
  df_samples <- data.frame(
    age = chronologicalAges,
    # Each sample's HR = exp( beta1 + beta3 * that person's age )
    hr  = exp( beta1 + beta3 * chronologicalAges )
  )
  
  # --------------------------------------------
  # 6. Plot: line for HR(Age), ribbon for 95% CI,
  #    dashed horizontal at HR=1, plus points at sample ages.
  # --------------------------------------------
  
  ggplot(df_plot, aes(x=age, y=hr)) +
    # Shaded 95% confidence region
    geom_ribbon(aes(ymin=hr_low, ymax=hr_high), alpha=0.2) +
    
    # The fitted HR(Age) curve
    geom_line(size=1) +
    
    # Horizontal reference line at HR=1
    geom_hline(yintercept=1, linetype="dashed", color="red") +
    
    # Per-sample points at their age, on the curve
    # (They should land exactly on the line if your model is purely age-based.)
    geom_point(data=df_samples, aes(x=age, y=hr), 
               color="blue", alpha=0.6) +
    
    labs(
      title = title,
      subtitle = "HR(+1 in Delta Age) as a function of Chronological Age\n(95% pointwise CI)",
      x = "Chronological Age",
      y = "Hazard Ratio"
    ) +
    theme_minimal()
}
```

```{r, plotHazardRatioMultiple}
plotHazardRatiosMultiDelta <- function(coxOutput, survDF, 
                                       delta_age_values = c(1, 2, 4),
                                       title = "HR by Chronological Age and Delta Age",
                                       x = "chronological", y = "delta_age", usingInteractionTerm = TRUE) {
  
  # Create data frame with all combinations of delta_age and age
  
  # Only include ages within the middle 80%
  q <- quantile(survDF$chronological, probs = c(0.10, 0.90))
  chronological <- survDF$chronological[survDF$chronological >= q[1] & survDF$chronological <= q[2], ]
  
  # chronological = survDF$chronological[survDF$chronological < quantile(survDF$chronological)[4]]
  
  df_plot <- expand.grid(
    chronological = chronological,
    delta_age = delta_age_values
  )
  
  coefficients <- rownames(as.data.frame(coxOutput$coefficients))
  if ("chronological_duplicate" %in% coefficients){ df_plot$chronological_duplicate <- df_plot$chronological }

  print("Predicting...")
  df_plot_predictions <- predict(coxOutput, newdata = df_plot, type = "lp", se.fit = TRUE)
  
  # Compute logHR and CI for each combo
  df_plot <- 
    df_plot |> dplyr::mutate(
      HR = exp(df_plot_predictions$fit),
      seBeta = sqrt(df_plot_predictions$se.fit),
      logHR = df_plot_predictions$fit,
      HR_low = exp(logHR - 1.96 * seBeta),
      HR_high = exp(logHR + 1.96 * seBeta)
    )

  # return(df_plot)
  print("Plotting...")
  ggplot(df_plot, aes(x = chronological, y = HR, color = factor(delta_age))) +
  geom_line(size = 1) +
  # geom_ribbon(aes(ymin = HR_low, ymax = HR_high, fill = factor(delta_age)), alpha = 0.15, color = NA) +
  geom_hline(yintercept = 1, linetype = "dashed", color = "red") +
  labs(
    title = title,
    subtitle = "Hazard Ratio for Different Delta Age Levels by Chronological Age",
    x = "Chronological Age",
    y = "Hazard Ratio",
    color = "Delta Age",
    fill = "Delta Age"
  ) +
  # xlim(min_age, max_age) +
  # coord_cartesian(ylim = c(minY, maxY)) +
  theme_minimal()
}

```

### Get the features used to calculate survival and run coxPH using them and Age
```{r, SurvivalOnlyFeatures}
survivalOnlyFeatures <- function(survivalOutput){
  selected_vars <- rownames(coef(survivalOutput$ModelBuilding$model, s = "lambda.min"))[coef(survivalOutput$ModelBuilding$model, s = "lambda.min")[,1] != 0]
  survDF <- survivalOutput$ModelBuilding$originalMetaTrn
  survDF$surv <- Surv(time = survDF$Survival_Time, event = as.numeric(factor(survDF$vital_status)) - 1)
  cox_formula <- as.formula(paste("surv ~ Age +", paste(selected_vars, collapse = "+")))
  cox_model <- coxph(cox_formula, data = survDF)
  summary(cox_model)
  return(cox_model)
}
```

```{r, FeatureImportance}
library(vip)
featureImportance <- function(result){
  # Calculate permutation feature importance
  vip_data <- vi(result$elnet, method = "permute", target = "Age", train = meta_age_associated, metric = "MAE",
                 pred_wrapper = function(object, newdata) predict(object, newdata))
  
  # Plot the variable importance
  vip(vip_data, geom = "point", num_features = 20, horiz = TRUE) +
    labs(title = "Permutation Feature Importance for Elastic Net Model (Full DataSet)")
}

```